<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.2" />
<title>raffiot.untyped.resource API documentation</title>
<meta name="description" content="Resource management module.
Ensure that create resources are always nicely released after use." />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>raffiot.untyped.resource</code></h1>
</header>
<section id="section-intro">
<p>Resource management module.
Ensure that create resources are always nicely released after use.</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">&#34;&#34;&#34;
Resource management module.
Ensure that create resources are always nicely released after use.
&#34;&#34;&#34;

from __future__ import annotations
from collections import abc
from dataclasses import dataclass
from raffiot.untyped import result, io, _MatchError
from raffiot.untyped.io import IO
from raffiot.untyped.result import Result, Ok, Error, Panic
from concurrent.futures import Executor
from raffiot.untyped.__internal import *


@dataclass
class Resource:
    &#34;&#34;&#34;
    Essentially an IO-powered data structure that produces resources of type A,
    can fail with errors of type E and read a context of type R.
    &#34;&#34;&#34;

    create: None
    &#34;&#34;&#34;
    IO to create one resource along with the IO for releasing it.
    
    On success, this IO must produce a `Tuple[A, IO[R,Any,Any]`:
        - The first value of the tuple, of type `A`, is the created resource.
        - The second value of the tuple, of type `IO[R,Any,Any]`, is the IO that
            release the first value.
    
    For example:
        
        &gt;&gt;&gt; Resource(io.defer(lambda: open(&#34;file&#34;)).map(
        &gt;&gt;&gt;     lambda a: (a, io.defer(a.close))))
    &#34;&#34;&#34;

    def use(self, fun):
        &#34;&#34;&#34;
        Create a resource a:A and use it in fun.

        Once created, the resource a:A is guaranteed to be released (by running
        its releasing IO). The return value if the result of fun(a).

        This is the only way to use a resource.
        &#34;&#34;&#34;

        def safe_use(x):
            a, close = x
            try:
                return (
                    fun(a)
                    .attempt()
                    .flat_map(lambda r: close.attempt().then(io.from_result(r)))
                )
            except Exception as exception:
                return close.attempt().then(io.panic(exception))

        return self.create.flat_map(safe_use)

    def map(self, f):
        &#34;&#34;&#34;
        Transform the created resource with f if the creation is successful.
        Do nothing otherwise.
        &#34;&#34;&#34;

        def safe_map(x):
            a, close = x
            try:
                return io.pure((f(a), close))
            except Exception as exception:
                return close.attempt().then(io.panic(exception))

        return Resource(self.create.flat_map(safe_map))

    def flat_map(self, f):
        &#34;&#34;&#34;
        Chain two Resource.
        The resource created by the first one (self) can be used to create the second (f).
        &#34;&#34;&#34;

        def safe_flat_map_a(xa):
            a, close_a = xa
            try:

                def safe_flat_map_a2(xa2):
                    a2, close_a2 = xa2
                    return io.pure((a2, close_a.attempt().then(close_a2)))

                return (
                    f(a)
                    .create.attempt()
                    .flat_map(
                        lambda r: r.fold(
                            safe_flat_map_a2,
                            lambda e: close_a.attempt().then(io.error(e)),
                            lambda p: close_a.attempt().then(io.panic(p)),
                        )
                    )
                )
            except Exception as exception:
                return close_a.attempt().then(io.panic(exception))

        return Resource(self.create.flat_map(safe_flat_map_a))

    def then(self, rs):
        &#34;&#34;&#34;
        Chain two Resource.
        The resource created by the first one (self) is dropped.
        &#34;&#34;&#34;
        return self.flat_map(lambda _: rs)

    def zip(self, *rs):
        &#34;&#34;&#34;
        Pack a list of resources (including self) into a Resource creating the
        list of all resources.

        If one resource creation fails, the whole creation fails (opened resources
        are released then).
        &#34;&#34;&#34;
        return zip((self, *rs))

    def ap(self, *arg):
        &#34;&#34;&#34;
        Noting functions from [X1,...,XN] to A: `[X1, ..., Xn] -&gt; A`.

        If self computes a function `f: [X1,...,XN] -&gt; A`
        and arg computes a value `x1: X1`,...,`xn: Xn`
        then self.ap(arg) computes `f(x1,...,xn): A`.
        &#34;&#34;&#34;
        return self.zip(*arg).map(lambda l: l[0](*l[1:]))

    def flatten(self):
        &#34;&#34;&#34;
        Concatenation function on Resource
        &#34;&#34;&#34;
        return self.flat_map(lambda x: x)

    # Reader API

    def contra_map_read(self, f):
        &#34;&#34;&#34;
        Transform the context with f.
        Note that f is not from R to R2 but from R2 to R!
        &#34;&#34;&#34;
        return Resource(
            self.create.contra_map_read(f).map(
                lambda x: (x[0], x[1].contra_map_read(f))
            )
        )

    # Error API

    def catch(self, handler):
        &#34;&#34;&#34;
        React to errors (the except part of a try-except).

        On error, call the handler with the error.
        &#34;&#34;&#34;
        return Resource(self.create.catch(lambda e: handler(e).create))

    def map_error(self, f):
        &#34;&#34;&#34;
        Transform the stored error if the resource creation fails on an error.
        Do nothing otherwise.
        &#34;&#34;&#34;
        return Resource(self.create.map_error(f))

    # Panic

    def recover(self, handler):
        &#34;&#34;&#34;
        React to panics (the except part of a try-except).

        On panic, call the handler with the exception.
        &#34;&#34;&#34;
        return Resource(self.create.recover(lambda p: handler(p).create))

    def map_panic(self, f):
        &#34;&#34;&#34;
        Transform the exception stored if the computation fails on a panic.
        Do nothing otherwise.
        &#34;&#34;&#34;
        return Resource(
            self.create.map_panic(f).map(lambda x: (x[0], x[1].map_panic(f)))
        )

    def attempt(self):
        &#34;&#34;&#34;
        Transform this Resource that may fail into a Resource
        that never fails but creates a Result.

        - If self successfully computes a, then `self.attempt()` successfully computes `Ok(a)`.
        - If self fails on error e, then `self.attempt()` successfully computes `Error(e)`.
        - If self fails on panic p, then `self.attempt()` successfully computes `Panic(p)`.

        Note that errors and panics stop the resource creation, unless a catch or
        recover reacts to such failures. But using map, flat_map, flatten and
        ap is sometimes easier than using catch and recover. attempt transforms
        a failed resource creation into a successful resource creation returning a failure,
        thus enabling you to use map, flat_map, ... to deal with errors.
        &#34;&#34;&#34;
        return Resource(
            self.create.attempt().map(
                lambda x: x.fold(
                    lambda v: (Ok(v[0]), v[1]),
                    lambda e: (Error(e), io.pure(None)),
                    lambda p: (Panic(p), io.pure(None)),
                )
            )
        )

    def finally_(self, rs):
        &#34;&#34;&#34;
        After having computed self, but before returning its result,
        execute the rs Resource creation.

        This is extremely useful when you need to perform an action,
        unconditionally, at the end of a resource creation, without changing
        its result, like executing a lifted IO.
        &#34;&#34;&#34;
        return self.attempt().flat_map(lambda r: rs.attempt().then(from_result(r)))

    def on_failure(self, handler):
        &#34;&#34;&#34;
        Combined form of catch and recover.
        React to any failure of the resource creation.
        Do nothing if the resource creation is successful.

        - The handler will be called on `Error(e)` if the resource creation fails with error e.
        - The handler will be called on `Panic(p)` if the resource creation fails with panic p.
        - The handler will never be called on `Ok(a)`.
        &#34;&#34;&#34;
        return self.attempt().flat_map(
            lambda r: r.fold(
                pure,
                lambda e: handler(result.error(e)),
                lambda p: handler(result.panic(p)),
            )
        )

    def contra_map_executor(self, f):
        &#34;&#34;&#34;
        Change the executor running this IO.
        :param f:
        :return:
        &#34;&#34;&#34;
        return Resource(self.create.contra_map_executor(f))


def liftIO(mio):
    &#34;&#34;&#34;
    Transform an IO into a Resource whose created resource if the result of the IO.
    The releasing IO does nothing.
    &#34;&#34;&#34;
    return Resource(mio.map(lambda a: (a, io.pure(None))))


def pure(a):
    &#34;&#34;&#34;
    A Resource that always returns the same constant.
    &#34;&#34;&#34;
    return Resource(io.pure((a, io.pure(None))))


def defer(deferred, *args, **kwargs):
    &#34;&#34;&#34;
    Defer a computation.

    The result of the Resource is the result of `deferred()`.

    For more details, see `io.defer`
    &#34;&#34;&#34;
    return liftIO(io.defer(deferred, *args, **kwargs))


def defer_resource(deferred, *args, **kwargs):
    &#34;&#34;&#34;
    Make a function that returns an `Resource`, a `Resource` itself.

    This is extremely useful with recursive function that would normally blow
    the stack (raise a stack overflow exception). Deferring recursive calls
    eliminates stack overflow.

    For more information see `io.defer_io`
    &#34;&#34;&#34;

    def f(*a, **kw):
        return deferred(*a, **kw).create

    return Resource(io.defer_io(f, *args, **kwargs))


def defer_read(deferred, *args, **kwargs):
    &#34;&#34;&#34;
    Similar to io.defer_read but for Resource.
    &#34;&#34;&#34;
    return liftIO(io.defer_read(deferred, *args, **kwargs))


def defer_read_resource(deferred, *args, **kwargs):
    &#34;&#34;&#34;
    Similar to io.defer_read_io but for Resource.
    &#34;&#34;&#34;

    def f(*a, **kw):
        return deferred(*a, **kw).create

    return Resource(io.defer_read_io(f, *args, **kwargs))


def read():
    &#34;&#34;&#34;
    Read the context.

    To execute a computation `IO[R,E,A]`, you need to call the run method with
    some value r of type R: `io.run(r)`. the `read()` action returns the value r
    given to run.

    Please note that the contra_map_read method can transform this value r.
    &#34;&#34;&#34;
    return liftIO(io.read())


def error(err):
    &#34;&#34;&#34;
    Resource creation that always fails on the error err.
    &#34;&#34;&#34;
    return Resource(io.error(err))


def panic(exception):
    &#34;&#34;&#34;
    Resource creation that always fails with the panic exception.
    &#34;&#34;&#34;
    return Resource(io.panic(exception))


def from_result(r):
    &#34;&#34;&#34;
    Resource creation that:
    - success if r is an `Ok`
    - fails with error e if r is `Error(e)`
    - fails with panic p if r is `Panic(p)`
    &#34;&#34;&#34;
    return r.fold(pure, error, panic)


def from_io_resource(mio):
    &#34;&#34;&#34;
    Construct a `Resource` from an `IO[R,E,Resource[R,E,A]]`
    &#34;&#34;&#34;
    return Resource(mio.flat_map(lambda rs: rs.create))


def from_open_close_io(open, close):
    &#34;&#34;&#34;
    Construct a `Resource` from an IO to open a resource and one to close it.
    &#34;&#34;&#34;
    return Resource(open.map(lambda a: (a, close(a))))


def from_open_close(open, close):
    &#34;&#34;&#34;
    Construct a `Resource` from a function to open a resource and one to close it.
    &#34;&#34;&#34;
    return Resource(io.defer(open).map(lambda a: (a, io.pure(a).map(close))))


def from_with(expr):
    &#34;&#34;&#34;
    Create a Resource from &#34;with&#34; python expression:

    &gt;&gt; with expr as x:
    &gt;&gt;     body

    is equivalent to

    &gt;&gt; from_with(lambda: expr)
    &#34;&#34;&#34;

    def manager_handler():
        manager = expr()
        enter = type(manager).__enter__
        exit = type(manager).__exit__
        (enter(manager), io.defer(lambda: exit(manager, None, None, None)))

    return Resource(io.defer(manager_handler))


def zip(*rs):
    &#34;&#34;&#34;
    Transform a list of Resource into a Resource creating a list
    of all resources.

    If once resource creation fails, the whole creation fails.
    The resources opened are then released.
    :param rs: the list of resources to pack.
    :return:
    &#34;&#34;&#34;
    if len(rs) == 1 and isinstance(rs[0], abc.Iterable):
        args = rs[0]
    else:
        args = rs

    def process(l):
        close = []
        args = []
        level = 0
        error = None

        for arg in l:
            if arg.is_ok():
                args.append(arg.success[0])
                close.append(arg.success[1])
            elif arg.is_error() and level &lt; 1:
                error = arg
                level = 1
            elif arg.is_panic() and level &lt; 2:
                error = arg
                level = 2
            else:
                level = 2
                error = Panic(_MatchError(f&#34;{arg} should be a Result &#34;))

        if level == 0:
            return io.pure((args, io.run_all(close)))
        else:
            return io.run_all(close).attempt().then(io.from_result(error))

    return Resource(io.zip([x.create.attempt() for x in args]).flat_map(process))


def traverse(l, f):
    &#34;&#34;&#34;
    Apply the function `f` to every element of the iterable.
    The resulting Resource creates the list of all the resources.

    This function is essentially like map, but f returns Resource[R,E,A2] instead of A2.

    :param l: the elements to apply to f
    :param f: the function for each element.
    :return:
    &#34;&#34;&#34;

    def append(l3):
        def do_it(a2):
            l3.append(a2)
            return l3

        return do_it

    r = pure([])

    def dumb(r, a):
        return r.flat_map(lambda l2: f(a).map(append(l2)))

    for a in l:
        r = dumb(r, a)
    return r


def yield_():
    &#34;&#34;&#34;
    Resource implement cooperative concurrency. It means a Resource has to
    explicitly make a break for other concurrent tasks to have a chance to
    progress.
    This is what `yeild_()` does, it forces the Resource to make a break,
    letting other tasks be run on the executor until the IO start progressing
    again.
    :return:
    &#34;&#34;&#34;
    return liftIO(io.yield_())


def async_(f):
    &#34;&#34;&#34;
    Perform an Asynchronous call. `f` is a function of the form:

    &gt;&gt;&gt; from concurrent.futures import Executor, Future
    &gt;&gt;&gt; def f(context: E,
    &gt;&gt;&gt;       executor: Executor,
    &gt;&gt;&gt;       callback: Callable[[Result], None]) -&gt; Future:
    &gt;&gt;&gt;     ...

    - `f` **MUST** return a `Future`.
    - `context` is the context of the Resource, usually the one passed to `run`
       if not changed by `contra_map_read`.
    - `executor` is the `Executor` where the Resource is run, usually the one
       passed to run if not changed by `contra_map_executor`.
    - `callback` **MUST ALWAYS BE CALLED EXACTLY ONCE**.
       Until `callback` is called, the Resource will be suspended waiting for the
       asynchronous call to complete.
       When `callback` is called, the Resource is resumed.
       The value passed to `callback` must be the result of the asynchonous call:

        - `Ok(value)` if the call was successful and returned `value`.
        - `Error(error)` if the call failed on error `error`.
        - `Panic(exception)` if the failed unexpectedly on exception `exception`.

    For example:

    &gt;&gt;&gt; from raffiot.result import Ok
    &gt;&gt;&gt; from raffiot.io import async_
    &gt;&gt;&gt; def f(context, executor, callback):
    &gt;&gt;&gt;     def h():
    &gt;&gt;&gt;         print(&#34;In the asynchronous call, returning 2.&#34;)
    &gt;&gt;&gt;         callback(Ok(2))
    &gt;&gt;&gt;     return executor.submit(h)
    &gt;&gt;&gt; resource = async_(f)
    &#34;&#34;&#34;
    return liftIO(io.async_(f))


def read_executor():
    &#34;&#34;&#34;
    Return the executor running this IO.
    :return:
    &#34;&#34;&#34;
    return liftIO(io.read_executor())</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="raffiot.untyped.resource.async_"><code class="name flex">
<span>def <span class="ident">async_</span></span>(<span>f)</span>
</code></dt>
<dd>
<div class="desc"><p>Perform an Asynchronous call. <code>f</code> is a function of the form:</p>
<pre><code class="language-python-repl">&gt;&gt;&gt; from concurrent.futures import Executor, Future
&gt;&gt;&gt; def f(context: E,
&gt;&gt;&gt;       executor: Executor,
&gt;&gt;&gt;       callback: Callable[[Result], None]) -&gt; Future:
&gt;&gt;&gt;     ...
</code></pre>
<ul>
<li><code>f</code> <strong>MUST</strong> return a <code>Future</code>.</li>
<li><code>context</code> is the context of the Resource, usually the one passed to <code>run</code>
if not changed by <code>contra_map_read</code>.</li>
<li><code>executor</code> is the <code>Executor</code> where the Resource is run, usually the one
passed to run if not changed by <code>contra_map_executor</code>.</li>
<li>
<p><code>callback</code> <strong>MUST ALWAYS BE CALLED EXACTLY ONCE</strong>.
Until <code>callback</code> is called, the Resource will be suspended waiting for the
asynchronous call to complete.
When <code>callback</code> is called, the Resource is resumed.
The value passed to <code>callback</code> must be the result of the asynchonous call:</p>
<ul>
<li><code>Ok(value)</code> if the call was successful and returned <code>value</code>.</li>
<li><code>Error(<a title="raffiot.untyped.resource.error" href="#raffiot.untyped.resource.error">error()</a>)</code> if the call failed on error <code><a title="raffiot.untyped.resource.error" href="#raffiot.untyped.resource.error">error()</a></code>.</li>
<li><code>Panic(exception)</code> if the failed unexpectedly on exception <code>exception</code>.</li>
</ul>
</li>
</ul>
<p>For example:</p>
<pre><code class="language-python-repl">&gt;&gt;&gt; from raffiot.result import Ok
&gt;&gt;&gt; from raffiot.io import async_
&gt;&gt;&gt; def f(context, executor, callback):
&gt;&gt;&gt;     def h():
&gt;&gt;&gt;         print(&quot;In the asynchronous call, returning 2.&quot;)
&gt;&gt;&gt;         callback(Ok(2))
&gt;&gt;&gt;     return executor.submit(h)
&gt;&gt;&gt; resource = async_(f)
</code></pre></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def async_(f):
    &#34;&#34;&#34;
    Perform an Asynchronous call. `f` is a function of the form:

    &gt;&gt;&gt; from concurrent.futures import Executor, Future
    &gt;&gt;&gt; def f(context: E,
    &gt;&gt;&gt;       executor: Executor,
    &gt;&gt;&gt;       callback: Callable[[Result], None]) -&gt; Future:
    &gt;&gt;&gt;     ...

    - `f` **MUST** return a `Future`.
    - `context` is the context of the Resource, usually the one passed to `run`
       if not changed by `contra_map_read`.
    - `executor` is the `Executor` where the Resource is run, usually the one
       passed to run if not changed by `contra_map_executor`.
    - `callback` **MUST ALWAYS BE CALLED EXACTLY ONCE**.
       Until `callback` is called, the Resource will be suspended waiting for the
       asynchronous call to complete.
       When `callback` is called, the Resource is resumed.
       The value passed to `callback` must be the result of the asynchonous call:

        - `Ok(value)` if the call was successful and returned `value`.
        - `Error(error)` if the call failed on error `error`.
        - `Panic(exception)` if the failed unexpectedly on exception `exception`.

    For example:

    &gt;&gt;&gt; from raffiot.result import Ok
    &gt;&gt;&gt; from raffiot.io import async_
    &gt;&gt;&gt; def f(context, executor, callback):
    &gt;&gt;&gt;     def h():
    &gt;&gt;&gt;         print(&#34;In the asynchronous call, returning 2.&#34;)
    &gt;&gt;&gt;         callback(Ok(2))
    &gt;&gt;&gt;     return executor.submit(h)
    &gt;&gt;&gt; resource = async_(f)
    &#34;&#34;&#34;
    return liftIO(io.async_(f))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.defer"><code class="name flex">
<span>def <span class="ident">defer</span></span>(<span>deferred, *args, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Defer a computation.</p>
<p>The result of the Resource is the result of <code>deferred()</code>.</p>
<p>For more details, see <code>io.defer</code></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def defer(deferred, *args, **kwargs):
    &#34;&#34;&#34;
    Defer a computation.

    The result of the Resource is the result of `deferred()`.

    For more details, see `io.defer`
    &#34;&#34;&#34;
    return liftIO(io.defer(deferred, *args, **kwargs))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.defer_read"><code class="name flex">
<span>def <span class="ident">defer_read</span></span>(<span>deferred, *args, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Similar to io.defer_read but for Resource.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def defer_read(deferred, *args, **kwargs):
    &#34;&#34;&#34;
    Similar to io.defer_read but for Resource.
    &#34;&#34;&#34;
    return liftIO(io.defer_read(deferred, *args, **kwargs))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.defer_read_resource"><code class="name flex">
<span>def <span class="ident">defer_read_resource</span></span>(<span>deferred, *args, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Similar to io.defer_read_io but for Resource.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def defer_read_resource(deferred, *args, **kwargs):
    &#34;&#34;&#34;
    Similar to io.defer_read_io but for Resource.
    &#34;&#34;&#34;

    def f(*a, **kw):
        return deferred(*a, **kw).create

    return Resource(io.defer_read_io(f, *args, **kwargs))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.defer_resource"><code class="name flex">
<span>def <span class="ident">defer_resource</span></span>(<span>deferred, *args, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Make a function that returns an <code><a title="raffiot.untyped.resource.Resource" href="#raffiot.untyped.resource.Resource">Resource</a></code>, a <code><a title="raffiot.untyped.resource.Resource" href="#raffiot.untyped.resource.Resource">Resource</a></code> itself.</p>
<p>This is extremely useful with recursive function that would normally blow
the stack (raise a stack overflow exception). Deferring recursive calls
eliminates stack overflow.</p>
<p>For more information see <code>io.defer_io</code></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def defer_resource(deferred, *args, **kwargs):
    &#34;&#34;&#34;
    Make a function that returns an `Resource`, a `Resource` itself.

    This is extremely useful with recursive function that would normally blow
    the stack (raise a stack overflow exception). Deferring recursive calls
    eliminates stack overflow.

    For more information see `io.defer_io`
    &#34;&#34;&#34;

    def f(*a, **kw):
        return deferred(*a, **kw).create

    return Resource(io.defer_io(f, *args, **kwargs))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.error"><code class="name flex">
<span>def <span class="ident">error</span></span>(<span>err)</span>
</code></dt>
<dd>
<div class="desc"><p>Resource creation that always fails on the error err.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def error(err):
    &#34;&#34;&#34;
    Resource creation that always fails on the error err.
    &#34;&#34;&#34;
    return Resource(io.error(err))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.from_io_resource"><code class="name flex">
<span>def <span class="ident">from_io_resource</span></span>(<span>mio)</span>
</code></dt>
<dd>
<div class="desc"><p>Construct a <code><a title="raffiot.untyped.resource.Resource" href="#raffiot.untyped.resource.Resource">Resource</a></code> from an <code>IO[R,E,<a title="raffiot.untyped.resource.Resource" href="#raffiot.untyped.resource.Resource">Resource</a>[R,E,A]]</code></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def from_io_resource(mio):
    &#34;&#34;&#34;
    Construct a `Resource` from an `IO[R,E,Resource[R,E,A]]`
    &#34;&#34;&#34;
    return Resource(mio.flat_map(lambda rs: rs.create))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.from_open_close"><code class="name flex">
<span>def <span class="ident">from_open_close</span></span>(<span>open, close)</span>
</code></dt>
<dd>
<div class="desc"><p>Construct a <code><a title="raffiot.untyped.resource.Resource" href="#raffiot.untyped.resource.Resource">Resource</a></code> from a function to open a resource and one to close it.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def from_open_close(open, close):
    &#34;&#34;&#34;
    Construct a `Resource` from a function to open a resource and one to close it.
    &#34;&#34;&#34;
    return Resource(io.defer(open).map(lambda a: (a, io.pure(a).map(close))))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.from_open_close_io"><code class="name flex">
<span>def <span class="ident">from_open_close_io</span></span>(<span>open, close)</span>
</code></dt>
<dd>
<div class="desc"><p>Construct a <code><a title="raffiot.untyped.resource.Resource" href="#raffiot.untyped.resource.Resource">Resource</a></code> from an IO to open a resource and one to close it.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def from_open_close_io(open, close):
    &#34;&#34;&#34;
    Construct a `Resource` from an IO to open a resource and one to close it.
    &#34;&#34;&#34;
    return Resource(open.map(lambda a: (a, close(a))))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.from_result"><code class="name flex">
<span>def <span class="ident">from_result</span></span>(<span>r)</span>
</code></dt>
<dd>
<div class="desc"><p>Resource creation that:
- success if r is an <code>Ok</code>
- fails with error e if r is <code>Error(e)</code>
- fails with panic p if r is <code>Panic(p)</code></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def from_result(r):
    &#34;&#34;&#34;
    Resource creation that:
    - success if r is an `Ok`
    - fails with error e if r is `Error(e)`
    - fails with panic p if r is `Panic(p)`
    &#34;&#34;&#34;
    return r.fold(pure, error, panic)</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.from_with"><code class="name flex">
<span>def <span class="ident">from_with</span></span>(<span>expr)</span>
</code></dt>
<dd>
<div class="desc"><p>Create a Resource from "with" python expression:</p>
<blockquote>
<blockquote>
<p>with expr as x:
body</p>
</blockquote>
</blockquote>
<p>is equivalent to</p>
<blockquote>
<blockquote>
<p>from_with(lambda: expr)</p>
</blockquote>
</blockquote></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def from_with(expr):
    &#34;&#34;&#34;
    Create a Resource from &#34;with&#34; python expression:

    &gt;&gt; with expr as x:
    &gt;&gt;     body

    is equivalent to

    &gt;&gt; from_with(lambda: expr)
    &#34;&#34;&#34;

    def manager_handler():
        manager = expr()
        enter = type(manager).__enter__
        exit = type(manager).__exit__
        (enter(manager), io.defer(lambda: exit(manager, None, None, None)))

    return Resource(io.defer(manager_handler))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.liftIO"><code class="name flex">
<span>def <span class="ident">liftIO</span></span>(<span>mio)</span>
</code></dt>
<dd>
<div class="desc"><p>Transform an IO into a Resource whose created resource if the result of the IO.
The releasing IO does nothing.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def liftIO(mio):
    &#34;&#34;&#34;
    Transform an IO into a Resource whose created resource if the result of the IO.
    The releasing IO does nothing.
    &#34;&#34;&#34;
    return Resource(mio.map(lambda a: (a, io.pure(None))))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.panic"><code class="name flex">
<span>def <span class="ident">panic</span></span>(<span>exception)</span>
</code></dt>
<dd>
<div class="desc"><p>Resource creation that always fails with the panic exception.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def panic(exception):
    &#34;&#34;&#34;
    Resource creation that always fails with the panic exception.
    &#34;&#34;&#34;
    return Resource(io.panic(exception))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.pure"><code class="name flex">
<span>def <span class="ident">pure</span></span>(<span>a)</span>
</code></dt>
<dd>
<div class="desc"><p>A Resource that always returns the same constant.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def pure(a):
    &#34;&#34;&#34;
    A Resource that always returns the same constant.
    &#34;&#34;&#34;
    return Resource(io.pure((a, io.pure(None))))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.read"><code class="name flex">
<span>def <span class="ident">read</span></span>(<span>)</span>
</code></dt>
<dd>
<div class="desc"><p>Read the context.</p>
<p>To execute a computation <code>IO[R,E,A]</code>, you need to call the run method with
some value r of type R: <code>io.run(r)</code>. the <code><a title="raffiot.untyped.resource.read" href="#raffiot.untyped.resource.read">read()</a></code> action returns the value r
given to run.</p>
<p>Please note that the contra_map_read method can transform this value r.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def read():
    &#34;&#34;&#34;
    Read the context.

    To execute a computation `IO[R,E,A]`, you need to call the run method with
    some value r of type R: `io.run(r)`. the `read()` action returns the value r
    given to run.

    Please note that the contra_map_read method can transform this value r.
    &#34;&#34;&#34;
    return liftIO(io.read())</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.read_executor"><code class="name flex">
<span>def <span class="ident">read_executor</span></span>(<span>)</span>
</code></dt>
<dd>
<div class="desc"><p>Return the executor running this IO.
:return:</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def read_executor():
    &#34;&#34;&#34;
    Return the executor running this IO.
    :return:
    &#34;&#34;&#34;
    return liftIO(io.read_executor())</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.traverse"><code class="name flex">
<span>def <span class="ident">traverse</span></span>(<span>l, f)</span>
</code></dt>
<dd>
<div class="desc"><p>Apply the function <code>f</code> to every element of the iterable.
The resulting Resource creates the list of all the resources.</p>
<p>This function is essentially like map, but f returns Resource[R,E,A2] instead of A2.</p>
<p>:param l: the elements to apply to f
:param f: the function for each element.
:return:</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def traverse(l, f):
    &#34;&#34;&#34;
    Apply the function `f` to every element of the iterable.
    The resulting Resource creates the list of all the resources.

    This function is essentially like map, but f returns Resource[R,E,A2] instead of A2.

    :param l: the elements to apply to f
    :param f: the function for each element.
    :return:
    &#34;&#34;&#34;

    def append(l3):
        def do_it(a2):
            l3.append(a2)
            return l3

        return do_it

    r = pure([])

    def dumb(r, a):
        return r.flat_map(lambda l2: f(a).map(append(l2)))

    for a in l:
        r = dumb(r, a)
    return r</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.yield_"><code class="name flex">
<span>def <span class="ident">yield_</span></span>(<span>)</span>
</code></dt>
<dd>
<div class="desc"><p>Resource implement cooperative concurrency. It means a Resource has to
explicitly make a break for other concurrent tasks to have a chance to
progress.
This is what <code>yeild_()</code> does, it forces the Resource to make a break,
letting other tasks be run on the executor until the IO start progressing
again.
:return:</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def yield_():
    &#34;&#34;&#34;
    Resource implement cooperative concurrency. It means a Resource has to
    explicitly make a break for other concurrent tasks to have a chance to
    progress.
    This is what `yeild_()` does, it forces the Resource to make a break,
    letting other tasks be run on the executor until the IO start progressing
    again.
    :return:
    &#34;&#34;&#34;
    return liftIO(io.yield_())</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.zip"><code class="name flex">
<span>def <span class="ident">zip</span></span>(<span>*rs)</span>
</code></dt>
<dd>
<div class="desc"><p>Transform a list of Resource into a Resource creating a list
of all resources.</p>
<p>If once resource creation fails, the whole creation fails.
The resources opened are then released.
:param rs: the list of resources to pack.
:return:</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def zip(*rs):
    &#34;&#34;&#34;
    Transform a list of Resource into a Resource creating a list
    of all resources.

    If once resource creation fails, the whole creation fails.
    The resources opened are then released.
    :param rs: the list of resources to pack.
    :return:
    &#34;&#34;&#34;
    if len(rs) == 1 and isinstance(rs[0], abc.Iterable):
        args = rs[0]
    else:
        args = rs

    def process(l):
        close = []
        args = []
        level = 0
        error = None

        for arg in l:
            if arg.is_ok():
                args.append(arg.success[0])
                close.append(arg.success[1])
            elif arg.is_error() and level &lt; 1:
                error = arg
                level = 1
            elif arg.is_panic() and level &lt; 2:
                error = arg
                level = 2
            else:
                level = 2
                error = Panic(_MatchError(f&#34;{arg} should be a Result &#34;))

        if level == 0:
            return io.pure((args, io.run_all(close)))
        else:
            return io.run_all(close).attempt().then(io.from_result(error))

    return Resource(io.zip([x.create.attempt() for x in args]).flat_map(process))</code></pre>
</details>
</dd>
</dl>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="raffiot.untyped.resource.Resource"><code class="flex name class">
<span>class <span class="ident">Resource</span></span>
<span>(</span><span>create: None)</span>
</code></dt>
<dd>
<div class="desc"><p>Essentially an IO-powered data structure that produces resources of type A,
can fail with errors of type E and read a context of type R.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@dataclass
class Resource:
    &#34;&#34;&#34;
    Essentially an IO-powered data structure that produces resources of type A,
    can fail with errors of type E and read a context of type R.
    &#34;&#34;&#34;

    create: None
    &#34;&#34;&#34;
    IO to create one resource along with the IO for releasing it.
    
    On success, this IO must produce a `Tuple[A, IO[R,Any,Any]`:
        - The first value of the tuple, of type `A`, is the created resource.
        - The second value of the tuple, of type `IO[R,Any,Any]`, is the IO that
            release the first value.
    
    For example:
        
        &gt;&gt;&gt; Resource(io.defer(lambda: open(&#34;file&#34;)).map(
        &gt;&gt;&gt;     lambda a: (a, io.defer(a.close))))
    &#34;&#34;&#34;

    def use(self, fun):
        &#34;&#34;&#34;
        Create a resource a:A and use it in fun.

        Once created, the resource a:A is guaranteed to be released (by running
        its releasing IO). The return value if the result of fun(a).

        This is the only way to use a resource.
        &#34;&#34;&#34;

        def safe_use(x):
            a, close = x
            try:
                return (
                    fun(a)
                    .attempt()
                    .flat_map(lambda r: close.attempt().then(io.from_result(r)))
                )
            except Exception as exception:
                return close.attempt().then(io.panic(exception))

        return self.create.flat_map(safe_use)

    def map(self, f):
        &#34;&#34;&#34;
        Transform the created resource with f if the creation is successful.
        Do nothing otherwise.
        &#34;&#34;&#34;

        def safe_map(x):
            a, close = x
            try:
                return io.pure((f(a), close))
            except Exception as exception:
                return close.attempt().then(io.panic(exception))

        return Resource(self.create.flat_map(safe_map))

    def flat_map(self, f):
        &#34;&#34;&#34;
        Chain two Resource.
        The resource created by the first one (self) can be used to create the second (f).
        &#34;&#34;&#34;

        def safe_flat_map_a(xa):
            a, close_a = xa
            try:

                def safe_flat_map_a2(xa2):
                    a2, close_a2 = xa2
                    return io.pure((a2, close_a.attempt().then(close_a2)))

                return (
                    f(a)
                    .create.attempt()
                    .flat_map(
                        lambda r: r.fold(
                            safe_flat_map_a2,
                            lambda e: close_a.attempt().then(io.error(e)),
                            lambda p: close_a.attempt().then(io.panic(p)),
                        )
                    )
                )
            except Exception as exception:
                return close_a.attempt().then(io.panic(exception))

        return Resource(self.create.flat_map(safe_flat_map_a))

    def then(self, rs):
        &#34;&#34;&#34;
        Chain two Resource.
        The resource created by the first one (self) is dropped.
        &#34;&#34;&#34;
        return self.flat_map(lambda _: rs)

    def zip(self, *rs):
        &#34;&#34;&#34;
        Pack a list of resources (including self) into a Resource creating the
        list of all resources.

        If one resource creation fails, the whole creation fails (opened resources
        are released then).
        &#34;&#34;&#34;
        return zip((self, *rs))

    def ap(self, *arg):
        &#34;&#34;&#34;
        Noting functions from [X1,...,XN] to A: `[X1, ..., Xn] -&gt; A`.

        If self computes a function `f: [X1,...,XN] -&gt; A`
        and arg computes a value `x1: X1`,...,`xn: Xn`
        then self.ap(arg) computes `f(x1,...,xn): A`.
        &#34;&#34;&#34;
        return self.zip(*arg).map(lambda l: l[0](*l[1:]))

    def flatten(self):
        &#34;&#34;&#34;
        Concatenation function on Resource
        &#34;&#34;&#34;
        return self.flat_map(lambda x: x)

    # Reader API

    def contra_map_read(self, f):
        &#34;&#34;&#34;
        Transform the context with f.
        Note that f is not from R to R2 but from R2 to R!
        &#34;&#34;&#34;
        return Resource(
            self.create.contra_map_read(f).map(
                lambda x: (x[0], x[1].contra_map_read(f))
            )
        )

    # Error API

    def catch(self, handler):
        &#34;&#34;&#34;
        React to errors (the except part of a try-except).

        On error, call the handler with the error.
        &#34;&#34;&#34;
        return Resource(self.create.catch(lambda e: handler(e).create))

    def map_error(self, f):
        &#34;&#34;&#34;
        Transform the stored error if the resource creation fails on an error.
        Do nothing otherwise.
        &#34;&#34;&#34;
        return Resource(self.create.map_error(f))

    # Panic

    def recover(self, handler):
        &#34;&#34;&#34;
        React to panics (the except part of a try-except).

        On panic, call the handler with the exception.
        &#34;&#34;&#34;
        return Resource(self.create.recover(lambda p: handler(p).create))

    def map_panic(self, f):
        &#34;&#34;&#34;
        Transform the exception stored if the computation fails on a panic.
        Do nothing otherwise.
        &#34;&#34;&#34;
        return Resource(
            self.create.map_panic(f).map(lambda x: (x[0], x[1].map_panic(f)))
        )

    def attempt(self):
        &#34;&#34;&#34;
        Transform this Resource that may fail into a Resource
        that never fails but creates a Result.

        - If self successfully computes a, then `self.attempt()` successfully computes `Ok(a)`.
        - If self fails on error e, then `self.attempt()` successfully computes `Error(e)`.
        - If self fails on panic p, then `self.attempt()` successfully computes `Panic(p)`.

        Note that errors and panics stop the resource creation, unless a catch or
        recover reacts to such failures. But using map, flat_map, flatten and
        ap is sometimes easier than using catch and recover. attempt transforms
        a failed resource creation into a successful resource creation returning a failure,
        thus enabling you to use map, flat_map, ... to deal with errors.
        &#34;&#34;&#34;
        return Resource(
            self.create.attempt().map(
                lambda x: x.fold(
                    lambda v: (Ok(v[0]), v[1]),
                    lambda e: (Error(e), io.pure(None)),
                    lambda p: (Panic(p), io.pure(None)),
                )
            )
        )

    def finally_(self, rs):
        &#34;&#34;&#34;
        After having computed self, but before returning its result,
        execute the rs Resource creation.

        This is extremely useful when you need to perform an action,
        unconditionally, at the end of a resource creation, without changing
        its result, like executing a lifted IO.
        &#34;&#34;&#34;
        return self.attempt().flat_map(lambda r: rs.attempt().then(from_result(r)))

    def on_failure(self, handler):
        &#34;&#34;&#34;
        Combined form of catch and recover.
        React to any failure of the resource creation.
        Do nothing if the resource creation is successful.

        - The handler will be called on `Error(e)` if the resource creation fails with error e.
        - The handler will be called on `Panic(p)` if the resource creation fails with panic p.
        - The handler will never be called on `Ok(a)`.
        &#34;&#34;&#34;
        return self.attempt().flat_map(
            lambda r: r.fold(
                pure,
                lambda e: handler(result.error(e)),
                lambda p: handler(result.panic(p)),
            )
        )

    def contra_map_executor(self, f):
        &#34;&#34;&#34;
        Change the executor running this IO.
        :param f:
        :return:
        &#34;&#34;&#34;
        return Resource(self.create.contra_map_executor(f))</code></pre>
</details>
<h3>Class variables</h3>
<dl>
<dt id="raffiot.untyped.resource.Resource.create"><code class="name">var <span class="ident">create</span> : NoneType</code></dt>
<dd>
<div class="desc"><p>IO to create one resource along with the IO for releasing it.</p>
<p>On success, this IO must produce a <code>Tuple[A, IO[R,Any,Any]</code>:
- The first value of the tuple, of type <code>A</code>, is the created resource.
- The second value of the tuple, of type <code>IO[R,Any,Any]</code>, is the IO that
release the first value.</p>
<p>For example:</p>
<pre><code>&gt;&gt;&gt; Resource(io.defer(lambda: open("file")).map(
&gt;&gt;&gt;     lambda a: (a, io.defer(a.close))))
</code></pre></div>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="raffiot.untyped.resource.Resource.ap"><code class="name flex">
<span>def <span class="ident">ap</span></span>(<span>self, *arg)</span>
</code></dt>
<dd>
<div class="desc"><p>Noting functions from [X1,&hellip;,XN] to A: <code>[X1, ..., Xn] -&gt; A</code>.</p>
<p>If self computes a function <code>f: [X1,...,XN] -&gt; A</code>
and arg computes a value <code>x1: X1</code>,&hellip;,<code>xn: Xn</code>
then self.ap(arg) computes <code>f(x1,...,xn): A</code>.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def ap(self, *arg):
    &#34;&#34;&#34;
    Noting functions from [X1,...,XN] to A: `[X1, ..., Xn] -&gt; A`.

    If self computes a function `f: [X1,...,XN] -&gt; A`
    and arg computes a value `x1: X1`,...,`xn: Xn`
    then self.ap(arg) computes `f(x1,...,xn): A`.
    &#34;&#34;&#34;
    return self.zip(*arg).map(lambda l: l[0](*l[1:]))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.attempt"><code class="name flex">
<span>def <span class="ident">attempt</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>Transform this Resource that may fail into a Resource
that never fails but creates a Result.</p>
<ul>
<li>If self successfully computes a, then <code>self.attempt()</code> successfully computes <code>Ok(a)</code>.</li>
<li>If self fails on error e, then <code>self.attempt()</code> successfully computes <code>Error(e)</code>.</li>
<li>If self fails on panic p, then <code>self.attempt()</code> successfully computes <code>Panic(p)</code>.</li>
</ul>
<p>Note that errors and panics stop the resource creation, unless a catch or
recover reacts to such failures. But using map, flat_map, flatten and
ap is sometimes easier than using catch and recover. attempt transforms
a failed resource creation into a successful resource creation returning a failure,
thus enabling you to use map, flat_map, &hellip; to deal with errors.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def attempt(self):
    &#34;&#34;&#34;
    Transform this Resource that may fail into a Resource
    that never fails but creates a Result.

    - If self successfully computes a, then `self.attempt()` successfully computes `Ok(a)`.
    - If self fails on error e, then `self.attempt()` successfully computes `Error(e)`.
    - If self fails on panic p, then `self.attempt()` successfully computes `Panic(p)`.

    Note that errors and panics stop the resource creation, unless a catch or
    recover reacts to such failures. But using map, flat_map, flatten and
    ap is sometimes easier than using catch and recover. attempt transforms
    a failed resource creation into a successful resource creation returning a failure,
    thus enabling you to use map, flat_map, ... to deal with errors.
    &#34;&#34;&#34;
    return Resource(
        self.create.attempt().map(
            lambda x: x.fold(
                lambda v: (Ok(v[0]), v[1]),
                lambda e: (Error(e), io.pure(None)),
                lambda p: (Panic(p), io.pure(None)),
            )
        )
    )</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.catch"><code class="name flex">
<span>def <span class="ident">catch</span></span>(<span>self, handler)</span>
</code></dt>
<dd>
<div class="desc"><p>React to errors (the except part of a try-except).</p>
<p>On error, call the handler with the error.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def catch(self, handler):
    &#34;&#34;&#34;
    React to errors (the except part of a try-except).

    On error, call the handler with the error.
    &#34;&#34;&#34;
    return Resource(self.create.catch(lambda e: handler(e).create))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.contra_map_executor"><code class="name flex">
<span>def <span class="ident">contra_map_executor</span></span>(<span>self, f)</span>
</code></dt>
<dd>
<div class="desc"><p>Change the executor running this IO.
:param f:
:return:</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def contra_map_executor(self, f):
    &#34;&#34;&#34;
    Change the executor running this IO.
    :param f:
    :return:
    &#34;&#34;&#34;
    return Resource(self.create.contra_map_executor(f))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.contra_map_read"><code class="name flex">
<span>def <span class="ident">contra_map_read</span></span>(<span>self, f)</span>
</code></dt>
<dd>
<div class="desc"><p>Transform the context with f.
Note that f is not from R to R2 but from R2 to R!</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def contra_map_read(self, f):
    &#34;&#34;&#34;
    Transform the context with f.
    Note that f is not from R to R2 but from R2 to R!
    &#34;&#34;&#34;
    return Resource(
        self.create.contra_map_read(f).map(
            lambda x: (x[0], x[1].contra_map_read(f))
        )
    )</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.finally_"><code class="name flex">
<span>def <span class="ident">finally_</span></span>(<span>self, rs)</span>
</code></dt>
<dd>
<div class="desc"><p>After having computed self, but before returning its result,
execute the rs Resource creation.</p>
<p>This is extremely useful when you need to perform an action,
unconditionally, at the end of a resource creation, without changing
its result, like executing a lifted IO.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def finally_(self, rs):
    &#34;&#34;&#34;
    After having computed self, but before returning its result,
    execute the rs Resource creation.

    This is extremely useful when you need to perform an action,
    unconditionally, at the end of a resource creation, without changing
    its result, like executing a lifted IO.
    &#34;&#34;&#34;
    return self.attempt().flat_map(lambda r: rs.attempt().then(from_result(r)))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.flat_map"><code class="name flex">
<span>def <span class="ident">flat_map</span></span>(<span>self, f)</span>
</code></dt>
<dd>
<div class="desc"><p>Chain two Resource.
The resource created by the first one (self) can be used to create the second (f).</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def flat_map(self, f):
    &#34;&#34;&#34;
    Chain two Resource.
    The resource created by the first one (self) can be used to create the second (f).
    &#34;&#34;&#34;

    def safe_flat_map_a(xa):
        a, close_a = xa
        try:

            def safe_flat_map_a2(xa2):
                a2, close_a2 = xa2
                return io.pure((a2, close_a.attempt().then(close_a2)))

            return (
                f(a)
                .create.attempt()
                .flat_map(
                    lambda r: r.fold(
                        safe_flat_map_a2,
                        lambda e: close_a.attempt().then(io.error(e)),
                        lambda p: close_a.attempt().then(io.panic(p)),
                    )
                )
            )
        except Exception as exception:
            return close_a.attempt().then(io.panic(exception))

    return Resource(self.create.flat_map(safe_flat_map_a))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.flatten"><code class="name flex">
<span>def <span class="ident">flatten</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>Concatenation function on Resource</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def flatten(self):
    &#34;&#34;&#34;
    Concatenation function on Resource
    &#34;&#34;&#34;
    return self.flat_map(lambda x: x)</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.map"><code class="name flex">
<span>def <span class="ident">map</span></span>(<span>self, f)</span>
</code></dt>
<dd>
<div class="desc"><p>Transform the created resource with f if the creation is successful.
Do nothing otherwise.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def map(self, f):
    &#34;&#34;&#34;
    Transform the created resource with f if the creation is successful.
    Do nothing otherwise.
    &#34;&#34;&#34;

    def safe_map(x):
        a, close = x
        try:
            return io.pure((f(a), close))
        except Exception as exception:
            return close.attempt().then(io.panic(exception))

    return Resource(self.create.flat_map(safe_map))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.map_error"><code class="name flex">
<span>def <span class="ident">map_error</span></span>(<span>self, f)</span>
</code></dt>
<dd>
<div class="desc"><p>Transform the stored error if the resource creation fails on an error.
Do nothing otherwise.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def map_error(self, f):
    &#34;&#34;&#34;
    Transform the stored error if the resource creation fails on an error.
    Do nothing otherwise.
    &#34;&#34;&#34;
    return Resource(self.create.map_error(f))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.map_panic"><code class="name flex">
<span>def <span class="ident">map_panic</span></span>(<span>self, f)</span>
</code></dt>
<dd>
<div class="desc"><p>Transform the exception stored if the computation fails on a panic.
Do nothing otherwise.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def map_panic(self, f):
    &#34;&#34;&#34;
    Transform the exception stored if the computation fails on a panic.
    Do nothing otherwise.
    &#34;&#34;&#34;
    return Resource(
        self.create.map_panic(f).map(lambda x: (x[0], x[1].map_panic(f)))
    )</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.on_failure"><code class="name flex">
<span>def <span class="ident">on_failure</span></span>(<span>self, handler)</span>
</code></dt>
<dd>
<div class="desc"><p>Combined form of catch and recover.
React to any failure of the resource creation.
Do nothing if the resource creation is successful.</p>
<ul>
<li>The handler will be called on <code>Error(e)</code> if the resource creation fails with error e.</li>
<li>The handler will be called on <code>Panic(p)</code> if the resource creation fails with panic p.</li>
<li>The handler will never be called on <code>Ok(a)</code>.</li>
</ul></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def on_failure(self, handler):
    &#34;&#34;&#34;
    Combined form of catch and recover.
    React to any failure of the resource creation.
    Do nothing if the resource creation is successful.

    - The handler will be called on `Error(e)` if the resource creation fails with error e.
    - The handler will be called on `Panic(p)` if the resource creation fails with panic p.
    - The handler will never be called on `Ok(a)`.
    &#34;&#34;&#34;
    return self.attempt().flat_map(
        lambda r: r.fold(
            pure,
            lambda e: handler(result.error(e)),
            lambda p: handler(result.panic(p)),
        )
    )</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.recover"><code class="name flex">
<span>def <span class="ident">recover</span></span>(<span>self, handler)</span>
</code></dt>
<dd>
<div class="desc"><p>React to panics (the except part of a try-except).</p>
<p>On panic, call the handler with the exception.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def recover(self, handler):
    &#34;&#34;&#34;
    React to panics (the except part of a try-except).

    On panic, call the handler with the exception.
    &#34;&#34;&#34;
    return Resource(self.create.recover(lambda p: handler(p).create))</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.then"><code class="name flex">
<span>def <span class="ident">then</span></span>(<span>self, rs)</span>
</code></dt>
<dd>
<div class="desc"><p>Chain two Resource.
The resource created by the first one (self) is dropped.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def then(self, rs):
    &#34;&#34;&#34;
    Chain two Resource.
    The resource created by the first one (self) is dropped.
    &#34;&#34;&#34;
    return self.flat_map(lambda _: rs)</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.use"><code class="name flex">
<span>def <span class="ident">use</span></span>(<span>self, fun)</span>
</code></dt>
<dd>
<div class="desc"><p>Create a resource a:A and use it in fun.</p>
<p>Once created, the resource a:A is guaranteed to be released (by running
its releasing IO). The return value if the result of fun(a).</p>
<p>This is the only way to use a resource.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def use(self, fun):
    &#34;&#34;&#34;
    Create a resource a:A and use it in fun.

    Once created, the resource a:A is guaranteed to be released (by running
    its releasing IO). The return value if the result of fun(a).

    This is the only way to use a resource.
    &#34;&#34;&#34;

    def safe_use(x):
        a, close = x
        try:
            return (
                fun(a)
                .attempt()
                .flat_map(lambda r: close.attempt().then(io.from_result(r)))
            )
        except Exception as exception:
            return close.attempt().then(io.panic(exception))

    return self.create.flat_map(safe_use)</code></pre>
</details>
</dd>
<dt id="raffiot.untyped.resource.Resource.zip"><code class="name flex">
<span>def <span class="ident">zip</span></span>(<span>self, *rs)</span>
</code></dt>
<dd>
<div class="desc"><p>Pack a list of resources (including self) into a Resource creating the
list of all resources.</p>
<p>If one resource creation fails, the whole creation fails (opened resources
are released then).</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def zip(self, *rs):
    &#34;&#34;&#34;
    Pack a list of resources (including self) into a Resource creating the
    list of all resources.

    If one resource creation fails, the whole creation fails (opened resources
    are released then).
    &#34;&#34;&#34;
    return zip((self, *rs))</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="raffiot.untyped" href="index.html">raffiot.untyped</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="two-column">
<li><code><a title="raffiot.untyped.resource.async_" href="#raffiot.untyped.resource.async_">async_</a></code></li>
<li><code><a title="raffiot.untyped.resource.defer" href="#raffiot.untyped.resource.defer">defer</a></code></li>
<li><code><a title="raffiot.untyped.resource.defer_read" href="#raffiot.untyped.resource.defer_read">defer_read</a></code></li>
<li><code><a title="raffiot.untyped.resource.defer_read_resource" href="#raffiot.untyped.resource.defer_read_resource">defer_read_resource</a></code></li>
<li><code><a title="raffiot.untyped.resource.defer_resource" href="#raffiot.untyped.resource.defer_resource">defer_resource</a></code></li>
<li><code><a title="raffiot.untyped.resource.error" href="#raffiot.untyped.resource.error">error</a></code></li>
<li><code><a title="raffiot.untyped.resource.from_io_resource" href="#raffiot.untyped.resource.from_io_resource">from_io_resource</a></code></li>
<li><code><a title="raffiot.untyped.resource.from_open_close" href="#raffiot.untyped.resource.from_open_close">from_open_close</a></code></li>
<li><code><a title="raffiot.untyped.resource.from_open_close_io" href="#raffiot.untyped.resource.from_open_close_io">from_open_close_io</a></code></li>
<li><code><a title="raffiot.untyped.resource.from_result" href="#raffiot.untyped.resource.from_result">from_result</a></code></li>
<li><code><a title="raffiot.untyped.resource.from_with" href="#raffiot.untyped.resource.from_with">from_with</a></code></li>
<li><code><a title="raffiot.untyped.resource.liftIO" href="#raffiot.untyped.resource.liftIO">liftIO</a></code></li>
<li><code><a title="raffiot.untyped.resource.panic" href="#raffiot.untyped.resource.panic">panic</a></code></li>
<li><code><a title="raffiot.untyped.resource.pure" href="#raffiot.untyped.resource.pure">pure</a></code></li>
<li><code><a title="raffiot.untyped.resource.read" href="#raffiot.untyped.resource.read">read</a></code></li>
<li><code><a title="raffiot.untyped.resource.read_executor" href="#raffiot.untyped.resource.read_executor">read_executor</a></code></li>
<li><code><a title="raffiot.untyped.resource.traverse" href="#raffiot.untyped.resource.traverse">traverse</a></code></li>
<li><code><a title="raffiot.untyped.resource.yield_" href="#raffiot.untyped.resource.yield_">yield_</a></code></li>
<li><code><a title="raffiot.untyped.resource.zip" href="#raffiot.untyped.resource.zip">zip</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="raffiot.untyped.resource.Resource" href="#raffiot.untyped.resource.Resource">Resource</a></code></h4>
<ul class="two-column">
<li><code><a title="raffiot.untyped.resource.Resource.ap" href="#raffiot.untyped.resource.Resource.ap">ap</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.attempt" href="#raffiot.untyped.resource.Resource.attempt">attempt</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.catch" href="#raffiot.untyped.resource.Resource.catch">catch</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.contra_map_executor" href="#raffiot.untyped.resource.Resource.contra_map_executor">contra_map_executor</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.contra_map_read" href="#raffiot.untyped.resource.Resource.contra_map_read">contra_map_read</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.create" href="#raffiot.untyped.resource.Resource.create">create</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.finally_" href="#raffiot.untyped.resource.Resource.finally_">finally_</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.flat_map" href="#raffiot.untyped.resource.Resource.flat_map">flat_map</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.flatten" href="#raffiot.untyped.resource.Resource.flatten">flatten</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.map" href="#raffiot.untyped.resource.Resource.map">map</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.map_error" href="#raffiot.untyped.resource.Resource.map_error">map_error</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.map_panic" href="#raffiot.untyped.resource.Resource.map_panic">map_panic</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.on_failure" href="#raffiot.untyped.resource.Resource.on_failure">on_failure</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.recover" href="#raffiot.untyped.resource.Resource.recover">recover</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.then" href="#raffiot.untyped.resource.Resource.then">then</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.use" href="#raffiot.untyped.resource.Resource.use">use</a></code></li>
<li><code><a title="raffiot.untyped.resource.Resource.zip" href="#raffiot.untyped.resource.Resource.zip">zip</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.2</a>.</p>
</footer>
</body>
</html>